## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [177]:
import pandas as pd


df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [178]:
# View the column names in the data
print(df.columns)


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [179]:
print(df.index)

RangeIndex(start=0, stop=54639, step=1)


In [180]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [181]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [182]:
df.isnull()

first   last    job  phone  email  client_id  order_id  order_date  \
0      False  False  False  False  False      False     False       False   
1      False  False  False  False  False      False     False       False   
2      False  False  False  False  False      False     False       False   
3      False  False  False  False  False      False     False       False   
4      False  False  False  False  False      False     False       False   
...      ...    ...    ...    ...    ...        ...       ...         ...   
54634  False  False  False  False  False      False     False       False   
54635  False  False  False  False  False      False     False       False   
54636  False  False  False  False  False      False     False       False   
54637  False  False  False  False  False      False     False       False   
54638  False  False  False  False  False      False     False       False   

       order_week  order_year  item_id  category  subcategory  unit_price  \
0           False       False    False     False        False       False   
1           False       False    False     False        False       False   
2           False       False    False     False        False       False   
3           False       False    False     False        False       False   
4           False       False    False     False        False       False   
...           ...         ...      ...       ...          ...         ...   
54634       False       False    False     False        False       False   
54635       False       False    False     False        False       False   
54636       False       False    False     False        False       False   
54637       False       False    False     False        False       False   
54638       False       False    False     False        False       False   

       unit_cost  unit_weight    qty  line_number  
0          False        False  False        False  
1          False        False  False        False  
2          False        False  False        False  
3          False        False  False        False  
4          False        False  False        False  
...          ...          ...    ...          ...  
54634      False        False  False        False  
54635      False        False  False        False  
54636      False        False  False        False  
54637      False        False  False        False  
54638      False        False  False        False  

[54639 rows x 18 columns]

In [183]:
# What three item categories had the most entries?
# non_null_counts = df.count()
# top_three_categories= non_null_counts.sort_values(ascending=False)
# top_three_categories.head(3)
df['category'].value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [184]:
# For the category with the most entries,
# which subcategory had the most entries?
# subcategory_with_most_entries = df['ca'].value_counts().idxmax()
# subcategory_with_most_entries
df.loc[df['category']== 'consumables','subcategory'].value_counts().head(1)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [185]:
# Which five clients had the most entries in the data?
# five_clients_with_most_entries = df['client_id'].value_counts()
# five_clients_with_most_entries.name = 'client_id'
# five_clients_with_most_entries.head(5)
df['client_id'].value_counts().head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [186]:
# Store the client ids of those top 5 clients in a list.
# five_clients_with_most_entries_to_df = pd.DataFrame(five_clients_with_most_entries)
# five_clients_with_most_entries_to_df.head(5).index.tolist()
list(df['client_id'].value_counts().head(5).index)

[33615, 66037, 46820, 38378, 24741]

In [187]:
# How many total units (the qty column) did the
# client with the most entries order order?
# top_client_id = df['client_id'].value_counts().idxmax()
# top_client_id
# total_units_ordered_by_client_with_most_entries = df[df['client_id'] == top_client_id]['qty'].sum()
# total_units_ordered_by_client_with_most_entries
df.loc[df['client_id']==33615, 'qty'].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [188]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['line_subtotal'] = df['unit_price']*df['qty']
# df.loc[df['unit_price','subcategory', 'line_subtotal' ]].head()
df['line_subtotal'].head()



0    115164.00
1       523.95
2       527.28
3      1056.18
4      3902.00
Name: line_subtotal, dtype: float64

In [189]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
def calculate_shipping_price(weight):
    if weight > 50:
        # $7 per pound for orders over 50 pounds
        return weight * 7
    else:
        # $10 per pound for items 50 pounds or under
        return weight * 10
df['shipping_price'] = df['unit_weight'].apply(calculate_shipping_price)*df['qty']
df['shipping_price'].head()


0    7875.0
1     312.9
2     655.2
3     356.7
4    9286.0
Name: shipping_price, dtype: float64

In [190]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
TAX_RATE = 0.0925
df['total_price'] = df['line_subtotal'] + df['shipping_price'] + (df['line_subtotal'] + df['shipping_price']) * TAX_RATE
df['total_price'].head()


0    134420.107500
1       914.258625
2      1291.859400
3      1543.571400
4     14407.890000
Name: total_price, dtype: float64

In [191]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['line_price'] = df['unit_price'] * df['qty'] + df['shipping_price']
df['line_price'].head()


0    123039.00
1       836.85
2      1182.48
3      1412.88
4     13188.00
Name: line_price, dtype: float64

In [204]:
# Create a column for the profit of
# each line using line cost and line price
df['profit'] = df['total_price'] - df['line_price']
df['profit'].sort_values(ascending=False).head()
#df['profit'].head()


21801    5.439340e+07
15112    2.036907e+07
25972    1.740790e+07
23929    1.417000e+07
12597    9.905656e+06
Name: profit, dtype: float64

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [205]:
# Check your work using the totals above
total_price_4_order_id_2742071 = df.loc[df['order_id'] == 2742071, 'total_price'].sum()
total_price_4_order_id_2742071

158105.649525

In [206]:
total_price_4_order_id_2173913 = df.loc[df['order_id'] == 2173913, 'total_price'].sum()
total_price_4_order_id_2173913

180488.05755

In [207]:
total_price_4_order_id_6128929 = df.loc[df['order_id'] == 6128929, 'total_price'].sum()
total_price_4_order_id_6128929

996593.762225

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [208]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
# This is to calculate the quantity purchased by each of the top 5 clients
top_clients_by_quantity = df.groupby('client_id')['qty'].sum().nlargest(5)
top_clients_by_quantity.head()

# top_clients_data.head(5)


client_id
78965    4033302
86721    1010426
30143     846069
85931     644423
77354     506382
Name: qty, dtype: int64

In [209]:
# This is to filter the original DataFrame to include only transactions from these top clients
top_clients_data = df[df['client_id'].isin(top_clients_by_quantity.index)]
top_clients_data.head(5)

first      last                                  job  \
332    Julie  Anderson                  Exhibition designer   
566  Randall    Watson                    Ceramics designer   
572   Dennis   Hawkins  Social research officer, government   
678      Amy   Stewart            Psychologist, educational   
745   Dennis   Hawkins  Social research officer, government   

                     phone                                email  client_id  \
332             5933099462        anderson_better_4704@bell.biz      78965   
566             3478635175  watson.hospital@scott-mcdonald.info      30143   
572  001-214-888-5838x7410          hawkins.dennis.5195@lee.com      85931   
678     979.793.4884x57205               stewartamy@francis.com      86721   
745  001-214-888-5838x7410          hawkins.dennis.5195@lee.com      85931   

     order_id  order_date  order_week  order_year  ... unit_price unit_cost  \
332   3928914  2023-04-21          16        2023  ...       2.39      1.90   
566   3348249  2023-05-03          18        2023  ...     655.42    466.88   
572   2407344  2023-02-06           6        2023  ...     206.59    175.46   
678   6057452  2023-03-21          12        2023  ...      45.68     34.27   
745   7438979  2023-03-27          13        2023  ...     234.04    123.86   

    unit_weight  qty  line_number  line_subtotal  shipping_price  \
332        4.76  485            0        1159.15         23086.0   
566        0.48   28            5       18351.76           134.4   
572       11.70   76            3       15700.84          8892.0   
678       10.37  339            2       15485.52         35154.3   
745        1.75   43            1       10063.72           752.5   

      total_price  line_price       profit  
332  26487.826375    24245.15  2242.676375  
566  20196.129800    18486.16  1709.969800  
572  26867.677700    24592.84  2274.837700  
678  55324.003350    50639.82  4684.183350  
745  11816.720350    10816.22  1000.500350  

[5 rows x 23 columns]

In [210]:
# This is calculate the total amount spent by quantity for each top client
amount_spent_by_client = top_clients_data.groupby('client_id')['qty'].sum().sort_values(ascending=False)
# This is to print out the top 5 results
print("Amount spent by each of the top 5 clients by quantity:")
amount_spent_by_client.head()

Amount spent by each of the top 5 clients by quantity:


client_id
78965    4033302
86721    1010426
30143     846069
85931     644423
77354     506382
Name: qty, dtype: int64

In [221]:
# This is to summarize the DataFrame showing the totals for the
# top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit.
summary_df = top_clients_data.groupby('client_id').agg(
    total_units_purchased=('qty', 'sum'),
    total_shipping_price=('shipping_price', 'sum'),
    total_revenue=('total_price', 'sum'),
    total_profit=('profit', 'sum')
).reset_index()
# This is to display thdefault top 5 results of the summary
summary_df.head()




client_id  total_units_purchased  total_shipping_price  total_revenue  \
0      30143                 846069            12830632.0   3.753358e+07   
1      77354                 506382            34020367.5   8.309236e+07   
2      78965                4033302             7626363.7   6.584207e+08   
3      85931                 644423            14459895.1   2.421556e+08   
4      86721                1010426            91905701.8   2.113865e+08   

   total_profit  
0  3.177901e+06  
1  7.035280e+06  
2  5.574729e+07  
3  2.050287e+07  
4  1.789771e+07

In [222]:
# This is to format the
# Currency in millions of dollars.
# Function
def to_millions(x):
    return '${:,.2f}M'.format(x / 1000000)
# Application of the function on the total revenue
summary_df['total_revenue'] = pd.to_numeric(summary_df['total_revenue'])
summary_df['total_revenue'] = summary_df['total_revenue'].apply(to_millions).sort_values(ascending=False)
summary_df['total_revenue'].head()

0     $37.53M
1     $83.09M
2    $658.42M
3    $242.16M
4    $211.39M
Name: total_revenue, dtype: object

In [224]:
# Application of the function on the total profit
summary_df['total_profit'] = summary_df['total_profit'].apply(to_millions).sort_values(ascending=False)
summary_df['total_profit'].head()

0     $3.18M
1     $7.04M
2    $55.75M
3    $20.50M
4    $17.90M
Name: total_profit, dtype: object

In [229]:
# This is to rename the columns
# to names suitable for presentation.
summary_df = summary_df.rename(columns={
    'client_id': 'Client ID',
    'total_units_purchased': 'Total Units Purchased',
    'total_shipping_price': 'Total Shipping Price',
    'total_revenue': 'Total Revenue (Millions)',
    'total_profit': 'Total Profit (Millions)'
})
# This is to display the summary datra frame after renaming columns 
# to names suitable for presentation
sorted_summary_df=summary_df.sort_values(by='Total Revenue (Millions)', ascending=False)
sorted_summary_df.head()

Client ID  Total Units Purchased  Total Shipping Price  \
1      77354                 506382            34020367.5   
2      78965                4033302             7626363.7   
0      30143                 846069            12830632.0   
3      85931                 644423            14459895.1   
4      86721                1010426            91905701.8   

  Total Revenue (Millions) Total Profit (Millions)  
1                  $83.09M                  $7.04M  
2                 $658.42M                 $55.75M  
0                  $37.53M                  $3.18M  
3                 $242.16M                 $20.50M  
4                 $211.39M                 $17.90M

In [230]:
# This is to sort the updated data by "Total Profit" from highest to lowest
summary_df.sort_values(by='Total Profit (Millions)', ascending=False)
summary_df.head()

Client ID  Total Units Purchased  Total Shipping Price  \
0      30143                 846069            12830632.0   
1      77354                 506382            34020367.5   
2      78965                4033302             7626363.7   
3      85931                 644423            14459895.1   
4      86721                1010426            91905701.8   

  Total Revenue (Millions) Total Profit (Millions)  
0                  $37.53M                  $3.18M  
1                  $83.09M                  $7.04M  
2                 $658.42M                 $55.75M  
3                 $242.16M                 $20.50M  
4                 $211.39M                 $17.90M

In [231]:
# Brief 2-3 sentence summary of my findings.
########################################
# While I believe my scripts are correct, 
# my outputs were not consistent with the expected 
# outputs. I have requested assistance from several 
# TAs who, while online atr the time of my assistance
# reuest, all rushed to set theior status to ""inactive"
# as soon as I told them I was still active and awaiting their support.
# Subsequently, no conclusion can be drawn art this moment until I
# get help to review where I might have gone wrong.
#######################################